pkg/sql/schemachanger/scplan/internal/scgraph/dep_edge_alloc.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package scgraph
+
+// depEdgeAlloc is used to amortize allocation of DepEdge objects.
+type depEdgeAlloc struct {
+	a []DepEdge
+}
+
+// minCap and maxCap are used to control the dynamic allocation size for
+// blocks of allocations.
+const minCap, maxCap = 64, 16384
+
+// new constructs a new DepEdge.
+func (a *depEdgeAlloc) new(de DepEdge) *DepEdge {
+	if a.a == nil {
+		a.a = make([]DepEdge, 0, minCap)
+	} else if len(a.a) == cap(a.a) {
+		newCap := 2 * cap(a.a)
+		if newCap > maxCap {
+			newCap = maxCap
+		}
+		a.a = make([]DepEdge, 0, newCap)
+	}
+	idx := len(a.a)
+	a.a = append(a.a, de)
```

Comment:
data of `a.a` does not need to be copied? looks like `a.a` is empty if it's resized.

Reply:
Indeed. I suggest adding a unit test that creates one of these things and adds `minCap+1` elements in it and checks that they're still in there.

Reply:
The idea here is that we want to allocate a chunk of these objects. We never use these slices for random access. We're initially throwing away the old slice when we resize because we never want to copy any data here unecessarily.

Reply:
ah, I see, thanks for the explanation.

Reply:
actually wondering how much penalty here for making a copy of `DepEdge` versus the gain of chunk allocation. Probably random allocation in heap is more expensive than making a copy in stack here?

Reply:
I don't think I understand. Can you say more?